test.js

```diff
@@ -0,0 +1,594 @@
+      t.is(server.connections.length, 0)
+      t.pass('server socket closed')
+    })
+
+    socket.on('error', (err) => t.fail('server socket error: ' + err.message + ' (' + err.code + ')'))
+  })
+
+  server.on('request', function (req, res) {
+    t.ok(req)
+    t.is(req.method, 'GET')
+    t.is(req.url, '/something/?key1=value1&key2=value2&enabled')
+    t.alike(req.headers, { host: server.address().address + ':' + server.address().port, connection: 'close' })
+    t.alike(req.getHeader('connection'), 'close')
+    t.alike(req.getHeader('Connection'), 'close')
+    t.ok(req.socket)
+    t.is(req.socket.requests.size, 1)
+
+    t.ok(res)
+    t.is(res.statusCode, 200, 'default status code')
+    t.alike(res.headers, {})
+    t.ok(res.socket)
+    t.is(res.request, req) // + should req also have req.response?
+    t.is(res.headersFlushed, false, 'headers not flushed')
+    t.is(res.chunked, true, 'chunked by default')
+    t.is(res.onlyHeaders, false)
+
+    t.is(req.socket, res.socket)
+    t.is(server.connections[req.socket.id], req.socket)
+
+    res.setHeader('Content-Length', 12)
+    t.alike(res.headers, { 'content-length': 12 })
+    t.is(res.getHeader('content-length'), 12)
+    t.is(res.getHeader('Content-Length'), 12)
+
+    res.end('Hello world!')
+
+    req.on('close', function () {
+      t.pass('server request closed')
+    })
+
+    res.on('close', function () {
+      t.is(res.headersFlushed, true, 'headers flushed')
+      t.is(res.chunked, false, 'not chunked')
+      t.pass('server response closed')
+    })
+  })
+
+  server.listen(0)
+  await waitForServer(server)
+
+  const req = http.request({
+    method: 'GET',
+    host: server.address().address,
+    port: server.address().port,
+    path: '/something/?key1=value1&key2=value2&enabled'
+  }, function (res) {
+    t.is(res.statusCode, 200)
+
+    const chunks = []
+    res.on('data', (chunk) => chunks.push(chunk))
+    res.on('end', () => {
+      const body = b4a.concat(chunks)
+      t.alike(body, b4a.from('Hello world!'), 'client response ended')
+    })
+  })
+
+  req.on('close', () => {
+    t.pass('request closed')
+
+    t.is(server.closing, false)
+    server.close()
+    t.is(server.closing, true)
+
+    server.on('close', function () {
+      t.pass('server closed')
+    })
+  })
+
+  req.on('error', (err) => t.fail('client req error: ' + err.message + ' (' + err.code + ')'))
+
+  req.end()
+})
+
+test('port already in use', async function (t) {
+  t.plan(2)
+
+  const server = createServer()
+  server.listen(0)
+  await waitForServer(server)
+
+  const server2 = createServer()
+  server2.listen(server.address().port)
+
+  server2.on('error', function (err) {
+    t.is(err.code, 'EADDRINUSE')
+
+    server.close()
+    server.on('close', () => t.pass('original server closed'))
+  })
+})
+
+// ../deps/uv/src/unix/core.c:178: uv_close: Assertion `0' failed
+// Aborted (core dumped)
+/* test.solo('destroy socket', async function (t) {
```

Comment:
This one also fails

Reply:
(btw I didn't finish writing this test but I left it as it is until the previous error is solved)